controllers/shooter.py

```diff
@@ -106,18 +118,38 @@ def tracking(self) -> None:
                 self.shooter.motor_speed = interpolate(
                     self.distance, self.ranges_lookup, self.flywheel_speed_lookup
                 )
-
-        if (
-            self._wants_to_fire
-            and self.indexer.has_cargo_in_chimney()
-            and self.shooter.is_at_speed()
-            and self.turret.is_on_target()
-            and self.distance > self.MIN_DIST
-            and self.distance < self.MAX_DIST
-            and self.chassis.translation_velocity.norm() < self.MAX_SPEED
-            and self.chassis.rotation_velocity.radians() < self.MAX_ROTATION
-        ):
-            self.next_state("firing")
+        accel = math.sqrt(
+            self.imu.getWorldLinearAccelX() ** 2 + self.imu.getWorldLinearAccelY() ** 2
+        )
+        if self.auto_shoot:
+            if (
+                self.indexer.has_cargo_in_chimney()
+                and self.shooter.is_at_speed()
+                and self.turret.is_on_target(
+                    math.atan(self.ALLOWABLE_TURRET_ERROR / self.distance)
+                )
+                and self.distance > self.AUTO_MIN_DIST
+                and self.distance < self.AUTO_MAX_DIST
+                and self.chassis.translation_velocity.norm() < self.AUTO_MAX_SPEED
+                and self.chassis.rotation_velocity.radians() < self.AUTO_MAX_ROTATION
+                and accel < self.AUTO_MAX_ACCEL
+            ):
+                self.next_state("firing")
+        else:
+            if (
+                self._wants_to_fire
+                and self.indexer.has_cargo_in_chimney()
+                and self.shooter.is_at_speed()
+                and self.turret.is_on_target(
+                    math.atan(self.AUTO_ALLOWABLE_TURRET_ERROR / self.distance)
+                )
+                and self.distance > self.MIN_DIST
+                and self.distance < self.MAX_DIST
+                and self.chassis.translation_velocity.norm() < self.MAX_SPEED
+                and self.chassis.rotation_velocity.radians() < self.MAX_ROTATION
+                and accel < self.MAX_ACCEL
+            ):
+                self.next_state("firing")
```

Comment:
Won't this mean that the driver cannot fire in autofire mode?

I think a better way (credit to Lucien here for this idea) is to have the autofire only take a shot when it is certain it can do it (ie tighter tolerance like you have here), but also respond to the driver trigger when the more lax human limits are met.